portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -25,7 +35,46 @@
 public class DataServlet extends HttpServlet {
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Joachim!</h1>");
+    Gson gson = new Gson();
+
+    ArrayList<String> data = new ArrayList<String>(0);
+    Query query = new Query("Comments");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      String comment = (String) entity.getProperty("comment");
+      data.add(comment);
+    }
+
+    String json = gson.toJson(data);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String comment = getComment(request);
+
+    Entity commentEntity = new Entity("Comments");
+    commentEntity.setProperty("comment", comment);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // data.add(0,comment);
+
+    datastore.put(commentEntity);
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+  }
+
+  private String getComment(HttpServletRequest request) {
+    String full_String = request.getParameter("name");
+    full_String += "~ ";
```

Comment:
My first attempt at adding this comment failed because I forgot to save it. 🤦‍♂ 

Instead of trying to serialize/deserialize comments using a string format like this, it would be better to create a more structured approach for storing comments.  You could introduce a new Comment class that takes care of moving between the formats you care about.

For example:

```
class Comment {
  private String author;
  private String comment;

  private Comment(String author, String comment) {
    this.author = author;
    this.comment = comment;
  }

  public static Comment fromEntity(Entity entity) {
    return new Comment(
        entity.getProperty(COMMENT_PROPERTY_AUTHOR),
        entity.getProperty(COMMENT_PROPERTY_COMMENT));
  }

  public static Comment fromServletRequest(HttpServletRequest request) {
    return new Comment(
        request.getParameter("name"),
        request.getParameter("comment-box"));
  }

  public Entity asEntity() {
    Entity entity = new Entity("Comments");
    entity.setProperty(COMMENT_PROPERTY_AUTHOR, author);
    entity.setProperty(COMMENT_PROPERTY_COMMENT, comment);
    return entity;
  }
}
```

This will encapsulate all of the information about the structure of the comments and how to convert them between different formats.

With this, if you wanted to create a comment from the servlet request and store it in the datastore, you can do the following:

```
Comment comment = Comment.fromServletRequest(request);
datastore.put(comment.asEntity());
```

Or when you're loading the comments from the datastore and returning them to the client, you can do the following:

```
commentList.add(Comment.fromEntity(entity));
...
gson.toJson(commentList);
```

And the Gson library will take care of converting the comment instances to valid JSON that looks like this:

```
[
  {"author": "Steve", "comment": "First!"},
  ... more comments ...
]
```